### PR TITLE
RUN-151: Fix: Properly detect when a job was renamed in SCM Import

### PR DESCRIPTION
**Pull request by ltamaster**
**Is this a bugfix, or an enhancement? Please describe.**
BigFix: when a job is renamed, SCM Import deletes and creates the job again

for issue https://github.com/rundeckpro/rundeckpro/issues/1731

**Describe the solution you've implemented**
- using jgit, check if a path was added and deleted (gitdiff)
- if that is the case save the renamed job in the import tracker
- when the job is imported, save the current UUID and the sourceId. Force to preserve original UUID when the job is re-imported.
- clean import trackers of the old path
- split job plugin metadata per integration

**Describe alternatives you've considered**

**Additional context**

### CODE REVIEW

**Comment by ltamaster** [comment]
@Mergifyio rebase

File: rundeckapp/grails-app/services/rundeck/services/scm/ScmLoaderService.groovy [hunk 1]
```diff
@@ -43,6 +42,7 @@ class ScmLoaderService implements EventBusAware {
     final Map<String, ScheduledFuture> scmProjectLoaderProcess = Collections.synchronizedMap([:])
     final Map<String, Boolean> scmProjectInitLoaded = Collections.synchronizedMap([:])
     final Map<String, ScmPluginConfigData> scmPluginMeta = Collections.synchronizedMap([:])
+    final Map<String, ScmPluginConfigData> scmFailedProjectInit = Collections.synchronizedMap([:])
```
Comment (gschueler) [inline rundeckapp/grails-app/services/rundeck/services/scm/ScmLoaderService.groovy:45]: I think these changes to the ScmLoaderService should not be part of this PR, it doesn't have to do with the renamed job import, right?

File: rundeckapp/grails-app/services/rundeck/services/ScmService.groovy [hunk 2]
```diff
@@ -1455,7 +1459,7 @@ class ScmService {
                 def newmeta = [name: job.getJobName(),groupPath: job.getGroupPath(),version: job.version, pluginMeta: jobStatus.commit.asMap()]
                 jobMetadataService.setJobPluginMeta(
                         job,
-                        STORAGE_NAME_IMPORT,
+                        STORAGE_NAME_EXPORT,
```
Comment (gschueler) [inline rundeckapp/grails-app/services/rundeck/services/ScmService.groovy:1462]: if the metadata now uses the integration type to store separate metadata for each job, how does it work when you have both import and export configured, and you e.g. rename a job?  previously the metadata was used to track the sourceId when you import a job, and later when you export it. is the sourceId still passed from import to export in the same way?

File: core/src/main/java/com/dtolabs/rundeck/plugins/scm/JobImporter.java [hunk 3]
```diff
@@ -62,10 +62,11 @@
      * @param input          input stream
      * @param importMetadata metadata to attach to the job
      * @param preserveUuid   if true, preserve any UUID on import, otherwise remove it
+     * @param renamedJob     pass original ID/UUID if the job is renamed
      *
      * @return result
      */
-    ImportResult importFromStream(String format, InputStream input, Map importMetadata, boolean preserveUuid);
+    ImportResult importFromStream(String format, InputStream input, Map importMetadata, boolean preserveUuid, Map renamedJob);
```
Comment (gschueler) [inline core/src/main/java/com/dtolabs/rundeck/plugins/scm/JobImporter.java:69]: looking at where this `renamedJob` is used, it would be better if it was not a Map, eg. add a new interface that has `getUuid()` and `getSourceId()`

**Review by gschueler — CHANGES_REQUESTED** [review]

**Comment by ltamaster** [comment]
@Mergifyio rebase

File: plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/GitImportPlugin.groovy [hunk 4]
```diff
@@ -436,6 +436,20 @@ class GitImportPlugin extends BaseGitPlugin implements ScmImportPlugin {
                         "${entry.changeType} ${entry.oldPath}->${entry.newPath}"
                     }.join("\n")
                     )
+
+                    String jobSerializerId = job?.sourceId !=null ? job.sourceId: job.id
+
+                    def findNewPath = changes.find { it.changeType == DiffEntry.ChangeType.ADD && it.newPath && it.newPath.contains(jobSerializerId) }
+                    def findOldPath = changes.find { it.changeType == DiffEntry.ChangeType.DELETE && it.oldPath && it.oldPath.contains(jobSerializerId) }
```
[the remark this answers is not in the document]
Comment (ltamaster) [reply plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/GitImportPlugin.groovy:443]: @gschueler 
when a job is renamed externally (it is an external event basically), the plugin cannot find the relation between the old path and the new path.
What I tried to do is to find  - according to the git changes - if the job was renamed

Screen Shot 2021-06-17 at 14 28 15

**Comment by ltamaster** [comment]
@Mergifyio rebase

File: rundeckapp/grails-app/services/rundeck/services/ScmService.groovy [hunk 5]
```diff
@@ -965,18 +941,24 @@ class ScmService {
      * @param serializer predefined serializer, or null to create lazy serializer
      * @return JobScmReference
      */
-    JobScmReference scmJobRef(ScheduledExecution job, JobSerializer serializer = null) {
-        def metadata = getJobPluginMeta(job)
-        scmJobRef(job, serializer, metadata)
+    JobScmReference scmImportJobRef(ScheduledExecution job, Map metadata = null) {
```
[the remark this answers is not in the document]
Comment (ltamaster) [reply rundeckapp/grails-app/services/rundeck/services/ScmService.groovy:944]: working on that

File: rundeckapp/grails-app/services/rundeck/services/scm/ScmLoaderService.groovy (hunk 1 above)
Comment (ltamaster) [reply rundeckapp/grails-app/services/rundeck/services/scm/ScmLoaderService.groovy:45]: on

File: rundeckapp/grails-app/services/rundeck/services/ScmService.groovy (hunk 2 above)
Comment (ltamaster) [reply rundeckapp/grails-app/services/rundeck/services/ScmService.groovy:1462]: @gschueler 
with this code
https://github.com/rundeck/rundeck/pull/7030/files#diff-8af69461e09e7c99b0829830e738be41d7d6809725ea047db37711357c3381acR440-R443

using the git changes plus the sourceID (or ID) I am trying to detect that the job was renamed.

File: plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/GitImportPlugin.groovy [hunk 6]
```diff
@@ -514,7 +528,14 @@ class GitImportPlugin extends BaseGitPlugin implements ScmImportPlugin {
...
             case JobChangeEvent.JobChangeEventType.MODIFY_RENAME:
-                importTracker.jobRenamed(reference, path, newpath)
+                if(importTracker.originalValue(newpath)){
+                    //if a job was imported with a new name traker needs to be cleaned
+                    def oldPath=importTracker.originalValue(newpath)
+                    importTracker.renamedTrackedItems.trackItem(oldPath, oldPath)
+                    importTracker.untrackPath(oldPath)
+                }else{
+                    importTracker.jobRenamed(reference, path, newpath)
+                }
```
[the remark this answers is not in the document]
Comment (ltamaster) [reply plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/GitImportPlugin.groovy:538]: I will add that

File: rundeckapp/grails-app/services/rundeck/services/ScmService.groovy (hunk 2 above)
Comment (gschueler) [reply rundeckapp/grails-app/services/rundeck/services/ScmService.groovy:1440]: here's the problem.  set up import using "archive" UUID behavior.  When imported, the uuid is removed, replaced with a new uuid, and the source uuid is stored in the job metadata (now under the "scm-import" name).  Then setup SCM export, using the export UUID behavior "original", this is meant to use the original source id in the export pattern, such as `${job.sourceId}.${config.format}`. However, the scm-export metadata does not contain the source Id anymore.  It used to work when the only metadata was stored with the "scm-import" metadata name.

**Review by gschueler — CHANGES_REQUESTED** [review]
* [ ] After external rename, before import, The scm import status still shows as "1 unimported files found, 1 tracked files were deleted"
* [x] The diff shows "import needed" and "file has been deleted from scm"
* [x] Performing the scm import works correctly. It no longer deletes/adds the job and maintains the execution history of the job.

After external rename status:
Screen Shot 2021-07-06 at 9 55 49 AM

diff:

Screen Shot 2021-07-06 at 9 56 18 AM

File: rundeckapp/grails-app/services/rundeck/services/ScmService.groovy (hunk 2 above)
Comment (ltamaster) [reply rundeckapp/grails-app/services/rundeck/services/ScmService.groovy:1514]: @gschueler I added more code to this case. The idea is to check (when the SCM export is enabled) if there is an SCM import metadata with a `sourceId`  attribute set

**Review by gschueler — APPROVED** [review]
- [x] project status shows job rename state. This seems to work now. 

I did see an instance of the same result where the tooltip for the project status showed "1 unimported file(s) found, 1 tracked file(s) were deleted", however when I refreshed the job list, it correctly showed "1 file was renamed" result, so I think it may have been resolving the status when I first loaded the page.

- [x] scm export correctly uses sourceId from original import now
